### PR TITLE
tools/crimson: stress tool update

### PR DESCRIPTION
**Pull request by ljx023**
modified some formatting and output names

change some experimental parameters

add a thread to detect version

output csv file

change README

### CODE REVIEW

File: tools/crimson/crimson_stress_tool.py [hunk 1]
```diff
@@ -443,6 +443,41 @@ def analyse(self):
         return result_dic
 
 
+class VersionThread(Task):
```
Comment (aisakaki) [inline tools/crimson/crimson_stress_tool.py:446]: This function should be better realized in general preprocessing

File: tools/crimson/crimson_stress_tool.py [hunk 2]
```diff
@@ -481,58 +487,74 @@ def run(self):
...
                 env.client_num = client_num
                 env.thread_num = thread_num
-                env.before_run_case()            
+                env.before_run_case()
                 tester = Tester(env)
-                test_case_result = tester.run()
+                temp_result = tester.run()
+                test_case_result = env.base_result.copy()
+                test_case_result['Thread_num'] = thread_num
+                test_case_result['Client_num'] = client_num
```
Comment (aisakaki) [inline tools/crimson/crimson_stress_tool.py:547]: TesterExecutor should try not to change test_case_result of one case. So it's better not to move this to here

File: tools/crimson/README.rst [hunk 3]
```diff
@@ -122,16 +123,31 @@ Example of result:
...
+                   OPtype     Rand Write    Rand Write    Rand Write    Rand Write    Rand Write    Rand Write
+                      OSD        Classic       Classic       Classic       Classic       Classic       Classic
+                    Store      Bluestore     Bluestore     Bluestore     Bluestore     Bluestore     Bluestore
+               Thread_num              2             4             6             2             4             6
+               Client_num              4             4             4             8             8             8
+          CPU-Utilization          486.2         877.9        1033.4         866.4        1195.0        1287.8
+    Context-Switches(K/s)         15.826        14.309        13.224        13.971        12.691        15.039
+       CPU-Migrations(/s)          1.725         1.115         1.875         1.706         1.972         1.924
+         Page-Faults(K/s)          8.462         4.129         8.108         3.937         5.316         4.611
+              page-faults         205992        181517        419542        170819        318061        297184
```
Comment (cyx1231st) [inline tools/crimson/README.rst:138]: Duplicated?

File: tools/crimson/crimson_stress_tool.py [hunk 4]
```diff
@@ -393,38 +398,38 @@ def analyse(self):
...
-                result_dic['iostat_read'] = float(temp_lis[2]) # kB/s
-                result_dic['iostat_write'] = float(temp_lis[8]) # kB/s
-                result_dic['iostat_avgrq-sz'] = float(temp_lis[21]) # sector/io
-                result_dic['iostat_rawait'] = float(temp_lis[5]) # ms
-                result_dic['iostat_wawait'] = float(temp_lis[11]) # ms
+                result_dic['Device_IPS'] = float(temp_lis[1])
+                result_dic['Device_OPS'] = float(temp_lis[7]) 
+                result_dic['Device_Read(MB/s)'] = round(float(temp_lis[2])/1024, 3)  # MB per second
+                result_dic['Device_Write(MB/s)'] = round(float(temp_lis[8])/1024, 3)  # MB per second
+                # result_dic['Device_Avgrq-sz(KB)'] = float(temp_lis[21]) # sector per io
```
Comment (cyx1231st) [inline tools/crimson/crimson_stress_tool.py:420]: Why comment out this?

Comment (cyx1231st) [reply tools/crimson/crimson_stress_tool.py:420]: Could you add aqu-sz/avgqu-sz instead?

File: tools/crimson/crimson_stress_tool.py [hunk 5]
```diff
@@ -481,58 +487,74 @@ def run(self):
...
+                    for key in keys_list:
+                        if "Bandwidth" in key:
+                            test_case_result["Bandwidth(MB/s)"] = \
+                                test_case_result.pop(key)
+                        elif "Latency" in key:
+                            test_case_result["Latency(ms)"] = \
+                                test_case_result.pop(key)
+                        elif "IOPS" in key:
+                            test_case_result["IOPS"] = \
+                                test_case_result.pop(key)
```
Comment (cyx1231st) [inline tools/crimson/crimson_stress_tool.py:527]: I don't quite follow these special handlings.

File: tools/crimson/README.rst [hunk 6]
```diff
@@ -122,16 +123,31 @@ Example of result:
...
+               Device_OPS        8449.65      15260.43      20073.58      15092.17      23670.17      36388.63
+        Device_Read(MB/s)          0.168         0.164         0.164         0.164         0.164         0.164
+       Device_Write(MB/s)         33.031        59.635        78.436        58.978        92.486       156.954
+        Device_Rawait(ms)           0.13          0.11           0.1           0.1          0.09           0.1
+        Device_Wawait(ms)           0.02          0.02          0.02          0.02          0.02          0.02
+                 CPU_Freq            1.0           1.0           1.0         1.001         1.034         1.393
+                  Version       20220124      20220124      20220124      20220124      20220124      20220124
+          Bandwidth(MB/s)          33.06       59.6981        78.562       59.0608        92.652      141.1881
+                     IOPS         8461.0       15281.0       20110.0       15116.0       23714.0       36142.0
+              Latency(ms)          0.944        1.0445 1.19099999999 1.05462500000       1.34575 1.32500000000
```
Comment (cyx1231st) [inline tools/crimson/README.rst:153]: probably prefix them as Client_Bandwidth, Client_IOPS and Client_Latency ?

File: tools/crimson/README.rst [hunk 7]
```diff
@@ -122,16 +123,31 @@ Example of result:
...
+    Instruction_per_Cycle           0.63          0.71          0.72          0.72          0.69          0.63
+          Branches(M/sec)        141.816       167.353       165.263       164.854       174.805       249.305
+            Branch-Misses          4.55%         3.06%         2.71%         3.03%         2.56%         2.33%
+               Device_IPS           7.82          7.83          7.85          7.83          7.83          7.87
+               Device_OPS        8449.65      15260.43      20073.58      15092.17      23670.17      36388.63
+        Device_Read(MB/s)          0.168         0.164         0.164         0.164         0.164         0.164
+       Device_Write(MB/s)         33.031        59.635        78.436        58.978        92.486       156.954
+        Device_Rawait(ms)           0.13          0.11           0.1           0.1          0.09           0.1
+        Device_Wawait(ms)           0.02          0.02          0.02          0.02          0.02          0.02
+                 CPU_Freq            1.0           1.0           1.0         1.001         1.034         1.393
```
Comment (cyx1231st) [inline tools/crimson/README.rst:149]: CPU_Freq(GHz)

**Review by cyx1231st — APPROVED** [review]

**Comment by ljx023** [comment]
@cyx1231st stress tool and visualizer [pr262](https://github.com/ceph/cbt/pull/262) have been modified.

**Pull request merged**